code/modules/antagonists/changeling/cellular_emporium.dm

```diff
@@ -25,35 +26,36 @@
 /datum/cellular_emporium/ui_data(mob/user)
 	var/list/data = list()
 
-	var/can_readapt = changeling.canrespec
-	var/genetic_points_remaining = changeling.geneticpoints
-	var/absorbed_dna_count = changeling.absorbedcount
+	data["can_readapt"] = changeling.can_respec
+
+	var/genetic_points_remaining = changeling.genetic_points
+	var/absorbed_dna_count = changeling.absorbed_count
 	var/true_absorbs = changeling.trueabsorbs
 
-	data["can_readapt"] = can_readapt
 	data["genetic_points_remaining"] = genetic_points_remaining
 	data["absorbed_dna_count"] = absorbed_dna_count
 
 	var/list/abilities = list()
 
-	for(var/path in changeling.all_powers)
-		var/datum/action/changeling/ability = path
+	for(var/datum/action/changeling/ability_path as anything in changeling.all_powers)
+
+		var/dna_cost = initial(ability_path.dna_cost)
+		var/req_dna = initial(ability_path.req_dna)
+		var/req_absorbs = initial(ability_path.req_absorbs)
 
-		var/dna_cost = initial(ability.dna_cost)
 		if(dna_cost <= 0)
 			continue
 
-		var/list/AL = list()
-		AL["name"] = initial(ability.name)
-		AL["desc"] = initial(ability.desc)
-		AL["helptext"] = initial(ability.helptext)
-		AL["owned"] = changeling.has_sting(ability)
-		var/req_dna = initial(ability.req_dna)
-		var/req_absorbs = initial(ability.req_absorbs)
-		AL["dna_cost"] = dna_cost
-		AL["can_purchase"] = ((req_absorbs <= true_absorbs) && (req_dna <= absorbed_dna_count) && (dna_cost <= genetic_points_remaining))
+		var/list/ability_data = list()
+		ability_data["name"] = initial(ability_path.name)
+		ability_data["desc"] = initial(ability_path.desc)
+		ability_data["path"] = ability_path
+		ability_data["helptext"] = initial(ability_path.helptext)
+		ability_data["owned"] = is_path_in_list_of_types(ability_path, changeling.purchased_powers)
+		ability_data["dna_cost"] = dna_cost
+		ability_data["can_purchase"] = ((req_absorbs <= true_absorbs) && (req_dna <= absorbed_dna_count) && (dna_cost <= genetic_points_remaining))
```

Comment:
ohhh brother. That's a lot to put on one line.
Maybe just like, suck it up and make it an if check?